src/cache.cc

```diff
@@ -10,14 +10,20 @@
 
 namespace {
 
-std::string GetCachedBaseFileName(const std::string& cache_directory,
-                                  std::string source_file) {
-  assert(!cache_directory.empty());
-  std::replace(source_file.begin(), source_file.end(), '\\', '_');
-  std::replace(source_file.begin(), source_file.end(), '/', '_');
-  std::replace(source_file.begin(), source_file.end(), ':', '_');
-
-  return cache_directory + source_file;
+std::string GetCachedBaseFileName(Config* config,
+                                  const std::string& source_file,
+                                  bool create_dir = false) {
+  assert(!config->cacheDirectory.empty());
+  std::string cache_file;
+  size_t len = config->projectRoot.size();
+  if (source_file.substr(0, len) == config->projectRoot) {
```

Comment:
StartsWith? That does not allocate.

Reply:
just putting this here for reference, its a lot better as a standard aproach to `StartsWith` than `substr` is: http://en.cppreference.com/w/cpp/string/basic_string/compare

Reply:
FWIW I prefer StartsWith over compare because it is more explicit.

Reply:
oh me too, thats what i meant by "for reference": in another project where you dont have StartsWith avaliable, `compare` is the way to do it (or the way to implement `StartsWith`)

Reply:
Ah, gotcha :)

Reply:
Years later we can use C++20 http://en.cppreference.com/w/cpp/string/basic_string/starts_with